tests/tests_ut/az_ulib_ipc_ut/az_ulib_ipc_ut.c

```diff
@@ -0,0 +1,1467 @@
+ * AZ_ULIB_ILLEGAL_ARGUMENT_ERROR. */
+TEST_FUNCTION(az_ulib_ipc_unpublish_with_null_descriptor_failed) {
+  /// arrange
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_SUCCESS, az_ulib_ipc_init(&g_ipc));
+  umock_c_reset_all_calls();
+
+  /// act
+  AZ_ULIB_RESULT result = az_ulib_ipc_unpublish(NULL, AZ_ULIB_NO_WAIT);
+
+  /// assert
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_ILLEGAL_ARGUMENT_ERROR, result);
+  ASSERT_ARE_EQUAL(int, 0, g_count_lock);
+  ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+  /// cleanup
+  az_ulib_ipc_deinit();
+}
+
+/* If the provided descriptor was not published, the az_ulib_ipc_unpublish shall return
+ * AZ_ULIB_ILLEGAL_ARGUMENT_ERROR. */
+TEST_FUNCTION(az_ulib_ipc_unpublish_with_unknown_descriptor_failed) {
+  /// arrange
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_SUCCESS, az_ulib_ipc_init(&g_ipc));
+  umock_c_reset_all_calls();
+
+  STRICT_EXPECTED_CALL(az_pal_os_lock_acquire(IGNORED_PTR_ARG));
+  STRICT_EXPECTED_CALL(az_pal_os_lock_release(IGNORED_PTR_ARG));
+
+  /// act
+  AZ_ULIB_RESULT result = az_ulib_ipc_unpublish(&MY_INTERFACE_1_123, AZ_ULIB_NO_WAIT);
+
+  /// assert
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_NO_SUCH_ELEMENT_ERROR, result);
+  ASSERT_ARE_EQUAL(int, 0, g_count_lock);
+  ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+  /// cleanup
+  az_ulib_ipc_deinit();
+}
+
+/* If one of the method in the interface is running and the wait policy is AZ_ULIB_NO_WAIT, the
+ * az_ulib_ipc_unpublish shall return AZ_ULIB_BUSY_ERROR. */
+TEST_FUNCTION(az_ulib_ipc_unpublish_with_method_running_failed) {
+  /// arrange
+  init_ipc_and_publish_interfaces();
+
+  my_method_model_in in;
+  in.action = 1;
+  in.descriptor = &MY_INTERFACE_1_123;
+  AZ_ULIB_RESULT out;
+
+  az_ulib_ipc_interface_handle interface_handle;
+  ASSERT_ARE_EQUAL(
+      int,
+      AZ_ULIB_SUCCESS,
+      az_ulib_ipc_try_get_interface(
+          MY_INTERFACE_1_123.name,
+          MY_INTERFACE_1_123.version,
+          AZ_ULIB_VERSION_EQUALS_TO,
+          &interface_handle));
+
+  umock_c_reset_all_calls();
+
+  STRICT_EXPECTED_CALL(az_pal_os_lock_acquire(IGNORED_PTR_ARG));
+  STRICT_EXPECTED_CALL(az_pal_os_lock_release(IGNORED_PTR_ARG));
+
+  /// act
+  // call umpublish inside of the methond.
+  AZ_ULIB_RESULT result = az_ulib_ipc_call(interface_handle, MY_METHOD_HANDLE, &in, &out);
+
+  /// assert
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_SUCCESS, result);
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_BUSY_ERROR, out);
+  ASSERT_ARE_EQUAL(int, 0, g_count_lock);
+  ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+  /// cleanup
+  az_ulib_ipc_release_interface(interface_handle);
+  az_ulib_ipc_unpublish(&MY_INTERFACE_1_2, AZ_ULIB_NO_WAIT);
+  az_ulib_ipc_unpublish(&MY_INTERFACE_3, AZ_ULIB_NO_WAIT);
+  az_ulib_ipc_unpublish(&MY_INTERFACE_1_123, AZ_ULIB_NO_WAIT);
+  az_ulib_ipc_unpublish(&MY_INTERFACE_2, AZ_ULIB_NO_WAIT);
+  az_ulib_ipc_deinit();
```

Comment:
nit: are you able to use `unpublish_interfaces_and_deinit_ipc()` here?

Reply:
Sure

---
In reply to: [338304248](https://github.com/Azure/azure-ulib-c/pull/40#discussion_r338304248) [](ancestors = 338304248)